pkg/services/ngalert/notifier/channels/template_data.go

```diff
@@ -0,0 +1,139 @@
+package channels
+
+import (
+	"net/url"
+	"path"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/common/model"
+)
+
+type ExtendedAlert struct {
+	Status       string      `json:"status"`
+	Labels       template.KV `json:"labels"`
+	Annotations  template.KV `json:"annotations"`
+	StartsAt     time.Time   `json:"startsAt"`
+	EndsAt       time.Time   `json:"endsAt"`
+	GeneratorURL string      `json:"generatorURL"`
+	Fingerprint  string      `json:"fingerprint"`
+	SilenceURL   string      `json:"silenceURL"`
+	DashboardURL string      `json:"dashboardURL"`
+	PanelURL     string      `json:"panelURL"`
+}
+
+type ExtendedAlerts []ExtendedAlert
+
+type ExtendedData struct {
+	Receiver string         `json:"receiver"`
+	Status   string         `json:"status"`
+	Alerts   ExtendedAlerts `json:"alerts"`
+
+	GroupLabels       template.KV `json:"groupLabels"`
+	CommonLabels      template.KV `json:"commonLabels"`
+	CommonAnnotations template.KV `json:"commonAnnotations"`
+
+	ExternalURL string `json:"externalURL"`
+}
+
+func removePrivateItems(kv template.KV) template.KV {
+	for key := range kv {
+		if strings.HasPrefix(key, "__") && strings.HasSuffix(key, "__") {
+			kv = kv.Remove([]string{key})
+		}
+	}
+	return kv
+}
+
+func extendAlert(alert template.Alert, externalURL string) ExtendedAlert {
+	extended := ExtendedAlert{
+		Status:       alert.Status,
+		Labels:       alert.Labels,
+		Annotations:  alert.Annotations,
+		StartsAt:     alert.StartsAt,
+		EndsAt:       alert.EndsAt,
+		GeneratorURL: alert.GeneratorURL,
+		Fingerprint:  alert.Fingerprint,
+	}
+
+	// fill in some grafana-specific urls
+	if len(externalURL) > 0 {
+		dashboardUid := alert.Annotations["__dashboardUid__"]
+		if len(dashboardUid) > 0 {
+			extended.DashboardURL = path.Join(externalURL, "/d/", dashboardUid)
```

Comment:
For joining URL we need to do something like this commit https://github.com/grafana/grafana/commit/6f02d70bbb62b0555e8dc65b631929a38c378b89. (If you noticed in the tests, it has become `http:/localhost` instead of `http://localhost`).

In case you got this code from the channel code, that will be changed shortly :)

Reply:
Same below for PanelURL

Reply:
fixed!